### PR TITLE
support sandboxes for raptor tasks

### LINKED ISSUE #2802: `task_sandbox` broken for raptor tasks
I believe this is a known problem, but I don't see that it is explicitly tracked and is likely to cause confusion.

1. The `task_sandbox` property of a raptor task generated on a client has a constructed path that is never created or used in the execution environment.
2. File staging directives with unqualified paths (relative paths instead of URIs) do not resolve to the directory in which raptor tasks actually run.
3. The `task:///` URI scheme does not work in staging directives raptor tasks.

### PR DESCRIPTION
**Pull request by andre-merzky**
This fixes #2802 

Test will be added.

### CODE REVIEW

File: src/radical/pilot/raptor/worker_default.py [hunk 1]
```diff
@@ -279,18 +279,29 @@ def _worker_proc(res_lock):
                 os.environ['CUDA_VISIBLE_DEVICES'] = \
                              ','.join(str(i) for i in task['slots']['gpus'])
 
+            sbox = task['description'].get('sandbox')
+            if not sbox:
+                sbox = self._sbox + '/' + task['uid']
+                task['description']['sandbox'] = sbox
```
Comment (mtitov) [inline src/radical/pilot/raptor/worker_default.py:285]: we should follow `sandbox` description in `TaskDescription`
```
That directory *MUST* be relative to the pilot sandbox
```
`sbox` should be taken from `task['task_sandbox_path']` (https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/session.py#L919, https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/tmgr/scheduler/base.py#L349)
```python
if task['description'].get('sandbox'):
    sbox = task['task_sandbox_path']
else:
    sbox = self._sbox + '/' + task['uid']
```
`self._sbox` is referred to Worker sandbox, right?

File: src/radical/pilot/raptor/worker_mpi.py [hunk 2]
```diff
@@ -434,11 +434,19 @@ def run(self):
                 self._prof.prof('advance',    uid=uid, state=AGENT_EXECUTING)
                 self._prof.prof('task_start', uid=uid)
 
+                sbox = task['description'].get('sandbox')
+                if not sbox:
+                    sbox = self._sbox + '/' + task['uid']
+                    task['description']['sandbox'] = sbox
```
Comment (mtitov) [inline src/radical/pilot/raptor/worker_mpi.py:440]: similar

File: tests/unit_tests/test_raptor/test_worker.py [hunk 3]
```diff
@@ -111,5 +116,45 @@ def test_call(self, mocked_init, mocked_Logger):
         self.assertEqual(exc, (None, None))
 
 
+    # --------------------------------------------------------------------------
+    #
+    @mock.patch.object(DefaultWorker, '__init__', return_value=None)
+    @mock.patch('radical.utils.Logger')
+    def test_sandbox(self, mocked_init, mocked_Logger):
```
Comment (mtitov) [inline tests/unit_tests/test_raptor/test_worker.py:123]: `test_sandbox(self, mocked_Logger, mocked_init)`

File: tests/unit_tests/test_raptor/test_worker.py [hunk 4]
```diff
@@ -111,5 +116,45 @@ def test_call(self, mocked_init, mocked_Logger):
...
+        component.check_pwd = os.getcwd
+        component._prof = mock.Mock()
+        component._log  = mocked_Logger
+        task = {'uid'        : 'task.0000',
+                'description': {
+                        'function': 'check_pwd',
+                        'sandbox' : '/tmp',
+                        'args'    : [],
+                        'kwargs'  : {}}}
+        out, err, ret, val, exc = component._dispatch_func(task)
```
Comment (mtitov) [inline tests/unit_tests/test_raptor/test_worker.py:142]: we should test `component._dispatch()` method here,
and for that we can patch `mp.Process` with `side_effect` as some method  that will receive `_worker_proc` and executes it.

**Review by mtitov — COMMENTED** [review]
added several comments

File: src/radical/pilot/raptor/worker_default.py (hunk 1 above)
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_default.py:285]: > self._sbox is referred to Worker sandbox, right?

Indeed

> That directory *MUST* be relative to the pilot sandbox

Hmm, I don't think we actually need this limitation anymore - the handling of absolute paths should be fixed by now.  Well, apart from what I added in this PR :-/  So let me fix this...

Ack on the `session._get_task_sandbox` call, but we need to have that available on the agent side also for raptor submitted tasks.

Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_default.py:285]: This is fixed now.  A side effect though is that the default sboxes are relative to the pilot, not the worker.  Which is consistent thoufg with other task which is good.

File: src/radical/pilot/raptor/worker_mpi.py (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_mpi.py]: indeed.

File: tests/unit_tests/test_raptor/test_worker.py (hunk 3 above)
Comment (andre-merzky) [reply tests/unit_tests/test_raptor/test_worker.py:123]: Thanks, fixed.

File: tests/unit_tests/test_raptor/test_worker.py (hunk 4 above)
Comment (andre-merzky) [reply tests/unit_tests/test_raptor/test_worker.py:142]: @mtitov , would it be imprudent to ask if you can take a stab at this?

Comment (mtitov) [reply tests/unit_tests/test_raptor/test_worker.py:142]: not at all, ok, will take car of it

Comment (andre-merzky) [reply tests/unit_tests/test_raptor/test_worker.py:142]: FWIW, another option would be to make `_dispatch` a wrapper around, say, dispatch_2 like this:
```py
def _dispatch(self, ...):
    return self._dispatch_2(...)
```
which would make `_dispatch_2` testable w/o the need for side effects.  Would that simplify things IYO?

Comment (mtitov) [reply tests/unit_tests/test_raptor/test_worker.py]: tests are updated (and `sys.exit` is considered within tests, but a corresponding question above is still valid)

File: src/radical/pilot/raptor/worker_mpi.py [hunk 5]
```diff
@@ -463,6 +467,7 @@ def run(self):
                 finally:
                     # send task back to rank 0
                     # FIXME: task_exec_stop
+                    os.chdir(self._sbox)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_mpi.py:470]: This was fixed meanwhile

[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_mpi.py:470]: Argh bugger - I fixed the linting in the raptor api branch, not here :-(  
https://github.com/radical-cybertools/radical.pilot/pull/2892/commits/a847ea5033c664343a88cd9c782d3b393837b1ff#diff-b357dc34307f7472a9d3a39e34f4908a9fce886863158a6e5736d1c3aeac4783R393

Let me see how we handle that...

File: tests/unit_tests/test_raptor/test_worker.py (hunk 4 above)
Comment (andre-merzky) [reply tests/unit_tests/test_raptor/test_worker.py]: > is it expected that _dispatch always finishes with sys.exit?

That is indeed the case for the `DefaultWorker`: it spawns a new process for each task and we use `sys.exit()` to reliably terminate that process again.  The `MPIWorker` though does not spawn new processes and thus does not  use `sys.exit`.

Comment (andre-merzky) [reply tests/unit_tests/test_raptor/test_worker.py]: Thanks for the test update!!

File: src/radical/pilot/raptor/worker_mpi.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/pilot/raptor/worker_mpi.py:471]: I pulled that one fix over manually now.  Sorry for fuddling this.

File: tests/unit_tests/test_raptor/test_worker.py (hunk 4 above)
Comment (mtitov) [reply tests/unit_tests/test_raptor/test_worker.py]: > > is it expected that _dispatch always finishes with sys.exit?
> 
> That is indeed the case for the `DefaultWorker`: it spawns a new process for each task and we use `sys.exit()` to reliably terminate that process again. The `MPIWorker` though does not spawn new processes and thus does not use `sys.exit`.

ah, I see, that makes sense, thank you!

**Review by mtitov — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**